stubs/requests/requests/sessions.pyi

```diff
@@ -45,8 +45,15 @@ class SessionRedirectMixin:
     def rebuild_proxies(self, prepared_request, proxies): ...
     def should_strip_auth(self, old_url, new_url): ...
 
-_Data: TypeAlias = str | bytes | Mapping[str, Any] | Iterable[tuple[str, str | None]] | IO[Any] | None
-
+_Auth: TypeAlias = Union[tuple[str, str], _auth.AuthBase, Callable[[PreparedRequest], PreparedRequest]]
+_Cert: TypeAlias = Union[str, tuple[str, str]]
+_Data: TypeAlias = str | bytes | Mapping[str, Any] | Iterable[tuple[str, str | None]] | IO[Any]
+_Files: TypeAlias = (
+    MutableMapping[str, IO[Any]]
+    | MutableMapping[str, tuple[str, IO[Any]]]
+    | MutableMapping[str, tuple[str, IO[Any], str]]
+    | MutableMapping[str, tuple[str, IO[Any], str, _TextMapping]]
+)
```

Comment:
Would you mind possibly providing some explanation as to how you figured out some of these more complex types? It might be quicker than us trawling through the `requests` codebase and repeating the steps you've already done :)

Also, just to check, were you looking at the `requests` `main` branch when you prepared this PR, or the 2.27 tag? If the former, note that the `main` branch is for `requests` 3.0, which hasn't yet been released -- these stubs are for `requests` 2.27. Some of the types look like they might be slightly different between 2.27 and `main`.

Reply:
Most of these types were already there, defined on `Session.request`: https://github.com/python/typeshed/blob/2299c43860c7115fabfd14ad854144e2ef20e132/stubs/requests/requests/sessions.pyi#L83
I only extracted them to a TypeAlias to reuse in get, post, put etc. (these all internally call `send`) and made some minor changes.

I didn't check them thoroughly, but it looks like they're mostly correct, especially if you follow [request's documentation](https://docs.python-requests.org/en/latest/api/).

Request's code seems to accept more than the documentation though, not sure how far we want to go in accepting arguments that would technically work, but are not documented.
For example:
- In `files`, the documentation specifically mentions to use 2/3/4-tuples, but passing lists also works.
- `allow_redirects` is documented as optional and `True` by default. Made the type `bool` now, but I think the code will accept `None`, and make it behave like `False`.

Reply:
Okay great, thanks, that's all really helpful!

Reply:
The diff for this PR is quite big, so I support keeping the annotations as they are for now and factoring them out into `TypeAlias`es, as you've done in this PR. We can always improve the types in a separate PR later on 👍

Reply:
For boolean parameters specifically, it's almost always the case that any object that supports `__bool__` (in other words, almost any object anywhere) can be passed and works fine at runtime. But that's not very useful for typing, so we do want to annotate such parameters as `bool` unless there's some specific reason (e.g., documentation) to allow a broader type.